### ISSUE TITLE
[1.16.5/1.17] Memory leaks in vanilla

### ISSUE THREAD

**Issue by BloCamLimb** [MinecraftForge/MinecraftForge#7819]
This is a bug of Minecraft Vanilla, I created an issue to Mojang, but it hasn't been fixed yet. I hope Forge does.

**Minecraft Version:** 1.16.5, 1.17

**Related Method:**
mcp: net.minecraft.client.renderer.texture.TextureUtil#readToBuffer(InputStream)
official: com.mojang.blaze3d.platform.TextureUtil#readResource(InputStream)

**Detailed Description:**
This method throws IOException, using **MemoryUtil.memAlloc()** for the buffer, and returns the pointer. A memory block allocated by **MemoryUtil.memAlloc()** must be freed by **MemoryUtil.memFree()** since it's not tracked by JVM or LWJGL. However, this method does not free the pointer. Although all usages of this method have a finally block containing **memFree()**. If an IOException occurred, this method will not run to the return statement, that means the argument passed to **memFree()** in the finally block is still null, but the memory block is allocated within this method. Apparently the memory leaks.

**Possible Solutions:**
Add a try catch block in this method catching any throwables, free the pointer first, then continue to throw it.
```
ByteBuffer ptr = null;
try {
    // reading
} catch (Throwable t) {
    memFree(ptr);
    throw t;
}
return ptr;
``` 
Or use BufferUtils from LWJGL for the buffer. It uses Java's Unsafe for native memory allocation, so it's tracked and can be garbage-collected by JVM.

**Comment by TheCurle** [comment]
Do you have statistics for the memory footprint this generates?
We're not in the business of fixing mojang bugs, but if this is a substantial problem we can ask them to put it on a higher priority.

**Comment by BloCamLimb** [comment]
No, I conclude from the code itself. This only happens when there's an error while reading textures or shaders.

**Comment by TheCurle** [comment]
If there is no measurable performance impact then we're not going to change anything. It's a trivial change, a maintenance nightmare, and unnecessary burden on our servers to create and store a new build that makes no difference for anyone.

If anyone in Forgecord is reading this, some testing would be appreciated :)

However, if it turns out that this leaks a few hundred MB of memory, then i'll even personally contribute the fix, because that's necessary.

**Comment by BloCamLimb** [comment]
It depends on how many bytes were read before the error occurred, and how many times the error occurred. Generally, there is no error reading, it won't make any difference. Because most of the errors occur when the file is not found, not when it is read. I think it's just a potential problem and not so serious.

**Comment by Codetoil** [comment]
Fixed as of 23w07a

**Comment by sciwhiz12** [comment]
Relevant Mojira issue is [MC-226729](https://bugs.mojang.com/browse/MC-226729).
